File: 2019/day01-rocket-equation/fuel.ts
#!/usr/bin/env ts-node
import { readAllLinesAsNumbers } from "@utils";
import * as _ from "lodash";

const sumFuelRequirements = (modules: number[], added: boolean) => _.sum(modules.map(added ? getAddedFuel : getFuel));
const getFuel = (mass: number) => Math.floor(mass / 3) - 2;
const getAddedFuel = (mass: number) => {
  let total = 0;
  do {
    mass = getFuel(mass);
    total += mass > 0 ? mass : 0;
  } while (mass > 0);
  return total;
};

if (require.main === module) {
  const modules: number[] = readAllLinesAsNumbers("day01.in");
  const [a, b] = ["Part 1", "Part 2"];
  console.time(a);
  console.timeLog(a, sumFuelRequirements(modules, false));
  console.time(b);
  console.timeLog(b, sumFuelRequirements(modules, true));
}

File: 2019/day01-rocket-equation/fuel.py
#!/usr/bin/env python


def getTotalFuel(masses: list) -> int:
    return sum([m//3 - 2 for m in masses])


def getMoreFuel(masses: list) -> int:
    return sum([getSum(m) for m in masses])


def getSum(mass: int) -> int:
    total = 0
    while True:
        mass = mass//3 - 2
        if mass <= 0:
            break
        total += mass
    return total


if __name__ == '__main__':
    with open('day01.in') as f:
        masses = [int(m.strip()) for m in f.readlines()]

    print(f"Part 1 = {getTotalFuel(masses)}")
    print(f"Part 2 = {getMoreFuel(masses)}")
